src/main/java/net/dv8tion/jda/internal/requests/restaction/GuildScheduledEventActionImpl.java

```diff
@@ -0,0 +1,201 @@
+        this.name = name;
+        return this;
+    }
+
+    @Nonnull
+    @Override
+    @CheckReturnValue
+    public GuildScheduledEventActionImpl setDescription(@Nullable String description)
+    {
+        if (description != null)
+            Checks.notLonger(description, GuildScheduledEvent.MAX_DESCRIPTION_LENGTH, "Description");
+        this.description = description;
+        return this;
+    }
+
+    @Nonnull
+    @Override
+    public GuildScheduledEventAction setImage(@Nullable Icon icon)
+    {
+        this.image = icon;
+        return this;
+    }
+
+    @Nonnull
+    @Override
+    public GuildScheduledEventAction setLocation(@Nonnull StageChannel stageChannel)
+    {
+        this.channelId = stageChannel.getIdLong();
+        this.entityType = 1;
+        return this;
+    }
+
+    @Nonnull
+    @Override
+    public GuildScheduledEventAction setLocation(@Nonnull VoiceChannel voiceChannel)
+    {
+        this.channelId = voiceChannel.getIdLong();
+        this.entityType = 2;
+        return this;
+    }
+
+    @Nonnull
+    @Override
+    public GuildScheduledEventAction setLocation(@Nonnull String externalLocation)
+    {
+        this.location = externalLocation;
+        this.entityType = 3;
+        return this;
+    }
+
+    @Nonnull
+    @Override
+    public GuildScheduledEventAction setStartTime(@Nonnull TemporalAccessor startTime)
+    {
+        Checks.notNull(startTime, "Temporal");
+        this.startTime = Helpers.toOffsetDateTime(startTime);
+        return this;
+    }
+
+    @Nonnull
+    @Override
+    public GuildScheduledEventAction setEndTime(@Nullable TemporalAccessor endTime)
+    {
+        Checks.notNull(endTime, "Temporal");
+        this.endTime = Helpers.toOffsetDateTime(endTime);
+        return this;
+    }
+
+    @Override
+    protected RequestBody finalizeData()
+    {
+        preChecks();
+        DataObject object = DataObject.empty();
+        object.put("entity_type", entityType);
+        object.put("privacy_level", 2);
+        object.put("name", name);
+        object.put("scheduled_start_time", startTime.format(DateTimeFormatter.ISO_DATE_TIME));
+        if (entityType == 1 || entityType == 2)
+            object.put("channel_id", channelId);
+        if (entityType == 3 && location != null && location.length() > 0)
+            object.put("entity_metadata", DataObject.empty().put("location", location));
+        if (description != null)
+            object.put("description", description);
+        if (image != null)
+            object.put("image", image.getEncoding());
+        if (endTime != null)
+            object.put("scheduled_end_time", endTime.format(DateTimeFormatter.ISO_DATE_TIME));
+
+        return getRequestBody(object);
+    }
+
+    private void preChecks()
+    {
+        Checks.check(name != null, "Missing required parameter: Name");
+        Checks.check(startTime != null, "Missing required parameter: Start Time");
```

Comment:
Required parameters should be part of the method to create instead of setters. Similar to all other createX functions in jda, required parameters are part of the method to get the rest action.

`createScheduledEvent(name, startTime, endTime, location)`

Reply:
done, still need to go over it a bit to find some inevitable oversights because it changes a lot of code and docs